src/cls/IPM/General/History.cls

```diff
@@ -0,0 +1,479 @@
+    Set ..Committed = (dev = 1) || $$$ISOK(status)
+    Set ..TimeEnd = $ZDateTime($Now(), 3)
+    If ..Package = "" {
+        Return $$$ERROR($$$GeneralError, "Package name not set")
+    }
+    Return ..%Save()
+}
+
+/// @API.Query
+/// Get load/install/uninstall records in the specified namespace
+Query Records(ascend As %Boolean, limit As %Integer, namespace As %String) As %Query(CONTAINID = 1, ROWSPEC = "Action:%String,Package:%String,Version_Major:%Integer,Version_Minor:%Integer,Version_Patch:%Integer,Version_Prerelease:%String,Version_Build:%String,SourceName:%String,SourceMoniker:%String,SourceDetails:%String,TimeStart:%String,TimeEnd:%String,NameSpace:%String,UserName:%String,Success:%String,Committed:%String,CommandString:%String") [ SqlProc ]
+{
+}
+
+ClassMethod RecordsExecute(ByRef qHandle As %Binary, ascend As %Boolean, limit As %Integer, namespace As %String) As %Status [ Internal ]
+{
+    Try {
+        Kill filter
+        Set qHandle = ..GetHistory(.filter, .ascend, .limit, .namespace)
+    } Catch e {
+        Return e.AsStatus()
+    }
+    Return $$$OK
+}
+
+ClassMethod RecordsFetch(ByRef qHandle As %Binary, ByRef Row As %List, ByRef AtEnd As %Integer = 0) As %Status [ Internal, PlaceAfter = RecordsExecute ]
+{
+    Try {
+        If qHandle.%Next() {
+            Set Row = $ListBuild(
+                qHandle.%Get("Action"),
+                qHandle.%Get("Package"),
+                qHandle.%Get("Version_Major"),
+                qHandle.%Get("Version_Minor"),
+                qHandle.%Get("Version_Patch"),
+                qHandle.%Get("Version_Prerelease"),
+                qHandle.%Get("Version_Build"),
+                qHandle.%Get("SourceName"),
+                qHandle.%Get("SourceMoniker"),
+                qHandle.%Get("SourceDetails"),
+                qHandle.%Get("TimeStart"),
+                qHandle.%Get("TimeEnd"),
+                qHandle.%Get("NameSpace"),
+                qHandle.%Get("UserName"),
+                qHandle.%Get("Success"),
+                qHandle.%Get("Committed"),
+                qHandle.%Get("CommandString")
+            )
+            Set AtEnd = 0
+        } Else {
+            Set Row = ""
+            Set AtEnd = 1
+        }
+    } Catch ex {
+        Return ex.AsStatus()
+    }
+    Return $$$OK
+}
+
+ClassMethod RecordsClose(ByRef qHandle As %Binary) As %Status [ Internal, PlaceAfter = RecordsFetch ]
+{
+    Set qHandle = ""
+    Quit $$$OK
+}
+
+/// Get the history of all installations, uninstalls, and loads in given namespace
+/// The filter argument is a multidimensional array with structure 
+/// filter(columnName) = value
+/// Where value can optionally start with >, >=, <, <=, =, <> or contain * to indicate a wildcard
+ClassMethod GetHistory(ByRef filter, ascend As %Boolean = 0, limit As %Integer = 0, namespace As %String) As %SQL.StatementResult
+{
+    New $namespace
+    If $Data(namespace) # 2 {
+        Set $namespace = $ZStrip(namespace, "<>WC")
```

Comment:
So over here, no namespace switching is needed. You can simply choose to include namespace filtering as an arg in the query to distinguish global vs local.
This is because the records are stored in a % global in %IPM.Storage.History so are in %SYS and available everywhere.
@isc-shuliu @isc-tleavitt is this the intention or should the history table be namespace specific, in which case this would need to switch to each namespace to run the query in?

Reply:
Just fixed it. We don't need the table to be namespace specific. All we care about is being able to filter by namespace.